src/include/parser/expression/abstract_expression.h

```diff
@@ -0,0 +1,122 @@
+#pragma once
+
+#include <functional>
+#include <memory>
+#include <utility>
+#include <vector>
+#include "common/hash_util.h"
+#include "common/json.h"
+#include "parser/expression/expression_defs.h"
+#include "type/type_id.h"
+#include "type/value.h"
+
+namespace terrier {
+namespace parser {
+namespace expression {
+
+/**
+ * An abstract parser expression. Dumb and immutable.
+ */
+class AbstractExpression {
+ protected:
+  /**
+   * Instantiates a new abstract expression. Because these are logical expressions, everything should be known
+   * at the time of instantiation, i.e. the resulting object is immutable.
+   * @param expression_type what type of expression we have
+   * @param return_value_type the type of the expression's value
+   * @param children the list of children for this node
+   */
+  AbstractExpression(const ExpressionType expression_type, const type::TypeId return_value_type,
+                     std::vector<std::unique_ptr<AbstractExpression>> &&children)
+      : expression_type_(expression_type), return_value_type_(return_value_type), children_(std::move(children)) {}
+
+  /**
+   * Copy constructs an abstract expression.
+   * @param other the abstract expression to be copied
+   */
+  AbstractExpression(const AbstractExpression &other)
+      : expression_type_(other.expression_type_), return_value_type_(other.return_value_type_) {
+    for (auto const &child : other.children_) {
+      children_.emplace_back(std::unique_ptr<AbstractExpression>(child->Copy()));
+    }
+  }
+
+ public:
+  virtual ~AbstractExpression() = default;
+
+  /**
+   * Hashes the current abstract expression.
+   */
+  virtual hash_t Hash() const {
+    hash_t hash = HashUtil::Hash(&expression_type_);
+    for (auto const &child : children_) {
+      hash = HashUtil::CombineHashes(hash, child->Hash());
+    }
+    return hash;
+  }
+
+  /**
+   * Logical equality check.
+   * @param rhs other
+   * @return true if the two expressions are logically equal
+   */
+  virtual bool operator==(const AbstractExpression &rhs) const {
+    if (expression_type_ != rhs.expression_type_ || children_.size() != rhs.children_.size()) {
+      return false;
+    }
+
+    for (size_t i = 0; i < children_.size(); i++) {
+      if (*children_[i].get() != *rhs.children_[i].get()) {
+        return false;
+      }
+    }
+
+    return true;
+  }
+
+  /**
+   * Logical inequality check.
+   * @param rhs other
+   * @return true if the two expressions are not logically equal
+   */
+  virtual bool operator!=(const AbstractExpression &rhs) const { return !(*this == rhs); }
+
+  /**
+   * Creates a copy of the current AbstractExpression.
+   */
+  virtual AbstractExpression *Copy() const { return new AbstractExpression(*this); }
```

Comment:
I find this abstraction weird since we use smart pointers most other places but raw pointers here.